Rule proposal: prefer-promise-strategies

raphinesse opened the pull request:
## Description

<!-- Describe your changes in detail. -->
This adds a new rule to enforce the use of the new `resolveTo` and `rejectWith` strategies instead of the more verbose alternative of using the `returnValue` strategy.

### Bad Code Example
```js
spy.withArgs(0).returnValue(Promise.resolve(123));
spy.and.returnValue(Promise.reject(123));
```

### Good Code Example
```js
spy.withArgs(0).resolveTo(123);
spy.and.rejectWith(123);
```

The rule also includes a fixer that will transform above bad into the good example.

### Motivation
Aside from leading to more concise code, transforming the `returnValue(Promise.reject())` example can also avoid an unwanted UnhandledPromiseRejectionWarning.

## How has this been tested?
I have added tests to cover my changes

## Types of changes

<!-- What types of changes does your code introduce? Put an `x` in all the boxes that apply: -->

- [ ] Test change (non-breaking change which adds additional test scenarios)
- [ ] Refactor change (non-breaking change updates coding styles)
- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

## Checklist

<!-- Go over all the following points, and put an `x` in all the boxes that apply. -->

- [x] I have ran `npm run test` and everything passes
- [x] My code follows the code style of this project
- [x] I have updated the documentation where necessary
- [x] I have read the [CONTRIBUTING](../CONTRIBUTING.md) document
- [x] I have added tests to cover my changes
- [x] All new and existing tests pass

raphinesse commented:
@DianaSuvorova In another issue you said that new rules were always a welcome addition. It would be great if you found the time to take a look at this proposal and let me know what you think.

DianaSuvorova commented on lib/rules/prefer-promise-strategies.js:
```diff
@@ -0,0 +1,47 @@
+
+module.exports = function (context) {
+  return {
+    // TODO: Change to "FOO:has(> BAR)" once available in esquery
```
If you are planning to leave this comment. Please add a link to esquery issue/ discussion/ anything that helps tracking the completion of this feature.

raphinesse replied:
Will do

raphinesse replied:
I removed it, since it does not really make a difference

DianaSuvorova commented on lib/rules/prefer-promise-strategies.js:
```diff
@@ -0,0 +1,47 @@
+const SpyStrategyCall = `
+  CallExpression:matches(
+    [callee.object.property.name=and],
+    [callee.object.callee.property.name=withArgs]
```
these probably will need an optional chaining or check..

raphinesse replied:
Since these chains are inside an ESQuery selector, they are safe AFAICT.

DianaSuvorova reviewed (commented):
Beautiful. Thank you so much for writing this out.  Yes, I think it is a very reasonable addition to the lib. Once you have test cases and the complete rule implementation  - I will be happy to take a look.
Fixer is especially appreciated.

Thank you!

raphinesse commented:
@DianaSuvorova Thanks for the quick reply. I will add the tests and docs and let you know then.

raphinesse commented:
@DianaSuvorova I rebased, integrated the rule and added tests and docs. I think this PR should be ready for a review.

DianaSuvorova reviewed (commented):
Thank you so much for the PR. I really like the way you add comments to your code. This is very helpful.  Left 2 comments. Otherwise looks great!

raphinesse commented on test/rules/prefer-promise-strategies.js:
```diff
@@ -0,0 +1,38 @@
+    { code: 'spy.and.returnValue(fn => fn(Promise.resolve(123)));' },
+    {
+      // Should be invalid but would need more complex analysis
+      code: 'const s = spy.and; s.returnValue(Promise.reject(123));'
```
Will do

raphinesse replied:
Done

raphinesse commented on lib/rules/prefer-promise-strategies.js:
```diff
@@ -0,0 +1,62 @@
+          end: returnStrategyCall.loc.end
+        },
+
+        fix (fixer) {
```
Isn't this covered by the expected `output` given in the `invalid` test cases? If not, could you kindly point me to an example fixer test?

DianaSuvorova replied:
Oh, yes, you are right.

raphinesse commented:
@DianaSuvorova I addressed your comments and made a few more improvements here and there. I put them in separate commits for an easier review, but they should be squashed before merging IMHO. Thanks a lot for your time and suggestions!

DianaSuvorova reviewed (approved):
Looks good! Thank you for your contribution!

tlvince commented:
:tada: This PR is included in version 4.1.3 :tada:

The release is available on:
- [npm package (@latest dist-tag)](https://www.npmjs.com/package/eslint-plugin-jasmine)
- [GitHub release](https://github.com/tlvince/eslint-plugin-jasmine/releases/tag/v4.1.3)

Your **[semantic-release](https://github.com/semantic-release/semantic-release)** bot :package::rocket:
